cpp/locality/NeighborQuery.h

```diff
@@ -81,13 +107,39 @@ class NeighborQuery
         //! Empty Destructor
         virtual ~NeighborQuery() {}
 
+        //! Perform a query based on a set of query parameters.
+        /*! Given a QueryArgs object and a set of points to perform a query
+         *  with, this function will dispatch the query to the appropriate
+         *  querying function.
+         *
+         *  This function should just be called query, but Cython's function
+         *  overloading abilities seem buggy at best, so it's easiest to just
+         *  rename the function.
+         */
+        virtual std::shared_ptr<NeighborQueryIterator> queryWithArgs(const vec3<float> *points, unsigned int N, QueryArgs args)
+            {
+            this->validateQueryArgs(args);
+            if (args.mode == QueryArgs::ball)
+                {
+                return this->queryBall(points, N, args.rmax, args.exclude_ii);
+                }
+            else if (args.mode == QueryArgs::nearest)
+                {
+                return this->query(points, N, args.nn, args.exclude_ii);
+                }
+            else
+                {
+                assert("Invalid query mode provided to generic query function.");
+                }
+            }
```

Comment:
What would you suggest? I left this warning because it's safe, but we can get rid of it by rewriting the code. One option would be to just assume that the function gets a valid argument. Another would be to rewrite the `else if` as just an `else`, and put in a nested `if` statement that checks that the value is the expected one and fails if that's the case. I don't want to throw an exception here since this function isn't in the public API, and any failure would be due to a developer error.

Reply:
@vyasr Maybe this warning would go away if you throw a runtime error in the `else` block?

Reply:
@vyasr I would still throw an exception. "Developer error" is a strange reason to reject this in my opinion - it's still an error that the user should know about so it can be debugged.

Reply:
After discussing offline, we came to the consensus that while this is an appropriate usage of assert, we want to move towards making it easier to eventually use freud as a pure C++ library. In that light, it makes sense to add more explicit error checking to C++ functions, so we can convert this to an exception.